frame/assets/src/lib.rs

```diff
@@ -247,19 +256,594 @@ decl_error! {
+		RefsLeft,
+		/// Invalid witness data given.
+		BadWitness,
 	}
 }
 
-decl_storage! {
-	trait Store for Module<T: Trait> as Assets {
-		/// The number of units of assets held by any given account.
-		Balances: map hasher(blake2_128_concat) (T::AssetId, T::AccountId) => T::Balance;
-		/// The next asset identifier up for grabs.
-		NextAssetId get(fn next_asset_id): T::AssetId;
-		/// The total unit supply of an asset.
+pub trait WeightInfo {
+	fn create() -> Weight;
+	fn force_create() -> Weight;
+	fn destroy(z: u32, ) -> Weight;
+	fn force_destroy(z: u32, ) -> Weight;
+	fn mint() -> Weight;
+	fn burn() -> Weight;
+	fn transfer() -> Weight;
+	fn force_transfer() -> Weight;
+	fn freeze() -> Weight;
+	fn thaw() -> Weight;
+	fn transfer_ownership() -> Weight;
+	fn set_team() -> Weight;
+	fn set_max_zombies() -> Weight;
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Issue a new class of fungible assets from a public origin.
+		///
+		/// This new asset class has no assets initially.
+		///
+		/// The origin must be Signed and the sender must have sufficient funds free.
+		///
+		/// Funds of sender are reserved according to the formula:
+		/// `AssetDepositBase + AssetDepositPerZombie * max_zombies`.
 		///
-		/// TWOX-NOTE: `AssetId` is trusted, so this is safe.
-		TotalSupply: map hasher(twox_64_concat) T::AssetId => T::Balance;
+		/// Parameters:
+		/// - `id`: The identifier of the new asset. This must not be currently in use to identify
+		/// an existing asset.
+		/// - `owner`: The owner of this class of assets. The owner has full superuser permissions
+		/// over this asset, but may later change and configure the permissions using `transfer_ownership`
+		/// and `set_team`.
+		/// - `max_zombies`: The total number of accounts which may hold assets in this class yet
+		/// have no existential deposit.
+		/// - `min_balance`: The minimum balance of this new asset that any single account must
+		/// have. If an account's balance is reduced below this, then it collapses to zero.
+		///
+		/// Emits `Created` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::create()]
+		fn create(origin,
+			#[compact] id: T::AssetId,
+			admin: <T::Lookup as StaticLookup>::Source,
+			max_zombies: u32,
+			min_balance: T::Balance,
+		) {
+			let owner = ensure_signed(origin)?;
+			let admin = T::Lookup::lookup(admin)?;
+
+			ensure!(!Asset::<T>::contains_key(id), Error::<T>::InUse);
+
+			let deposit = T::AssetDepositPerZombie::get()
+				.saturating_mul(max_zombies.into())
+			 	.saturating_add(T::AssetDepositBase::get());
+			T::Currency::reserve(&owner, deposit)?;
+
+			Asset::<T>::insert(id, AssetDetails {
+				owner: owner.clone(),
+				issuer: admin.clone(),
+				admin: admin.clone(),
+				freezer: admin.clone(),
+				supply: Zero::zero(),
+				deposit,
+				max_zombies,
+				min_balance,
```

Comment:
```suggestion
				min_balance.max(1),
```

Min balance of 0 would mess things up with regards to our accounting and destroying the token right?

Reply:
Maybe better to put an ensure statement so that users do not create a token with the wrong idea

Reply:
This test I created fails:

```
	#[test]
	fn min_balance_should_work_zero() {
		new_test_ext().execute_with(|| {
			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 0));
			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, 100));
			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 1);

			// When deducting from an account to below minimum, it should be reaped.

			assert_ok!(Assets::transfer(Origin::signed(1), 0, 2, 100));
			assert!(Assets::balance(0, 1).is_zero());
			assert_eq!(Assets::balance(0, 2), 100);
			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 1);

			assert_ok!(Assets::force_transfer(Origin::signed(1), 0, 2, 1, 100));
			assert!(Assets::balance(0, 2).is_zero());
			assert_eq!(Assets::balance(0, 1), 100);
			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 1);

			assert_ok!(Assets::burn(Origin::signed(1), 0, 1, 100));
			assert!(Assets::balance(0, 1).is_zero());
			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 0);
		});
	}
```

So at least the `burn` function would need an update in logic to handle `min_balance = 0`

Reply:
Will require `min_balance` > 0 for now.